### PR TITLE
Syntax operations also need to ensure project is present for the open script infos since update could be pending to make sure open script info has project

### LINKED ISSUE #50131: No Project error during getOutliningSpans
Present in 4.7.4. and `main`.

1. `git clone https://github.com/facebook/docusaurus`
2. `cd docusaurus`
3. `git reset --hard 9150c7aefc7bd871f449145bee5bc70767d09eca`
4. `cd ..`
5. `npm install @typescript/server-replay`
6. Download replay.txt
7. `npx tsreplay ./docusaurus path/to/replay.txt path/to/tsserver.js`

<details>
<summary>replay.txt</summary>

```json
{"rootDirPlaceholder":"@PROJECT_ROOT@","serverArgs":["--disableAutomaticTypingAcquisition"]}
{"seq":1,"type":"request","command":"configure","arguments":{"preferences":{"includePackageJsonAutoImports":"off"},"watchOptions":{"excludeDirectories":["**/node_modules"]}}}
{"seq":2,"type":"request","command":"updateOpen","arguments":{"changedFiles":[],"closedFiles":[],"openFiles":[{"file":"@PROJECT_ROOT@/packages/docusaurus-module-type-aliases/src/index.d.ts","projectRootPath":"@PROJECT_ROOT@"}]}}
{"seq":3,"type":"request","command":"updateOpen","arguments":{"changedFiles":[],"closedFiles":["@PROJECT_ROOT@/packages/docusaurus-module-type-aliases/src/index.d.ts"],"openFiles":[]}}
{"seq":4,"type":"request","command":"updateOpen","arguments":{"changedFiles":[],"closedFiles":[],"openFiles":[{"file":"@PROJECT_ROOT@/packages/docusaurus-plugin-client-redirects/src/__tests__/redirectValidation.test.ts","projectRootPath":"@PROJECT_ROOT@"}]}}
{"seq":5,"type":"request","command":"getOutliningSpans","arguments":{"file":"@PROJECT_ROOT@/packages/docusaurus-plugin-client-redirects/src/__tests__/redirectValidation.test.ts"}}
```

</details>

```
Error: No Project.
    at Object.ThrowNoProject (typescript/built/local/tsserver.js:167629:23)
    at IOSession.Session.getFileAndLanguageServiceForSyntacticOperation (typescript/built/local/tsserver.js:176275:42)
    at IOSession.Session.getOutliningSpans (typescript/built/local/tsserver.js:176288:31)
    at Session.handlers.ts.Map.ts.getEntries._a.<computed> (typescript/built/local/tsserver.js:175074:61)
    at typescript/built/local/tsserver.js:177179:88
    at IOSession.Session.executeWithRequestId (typescript/built/local/tsserver.js:177170:28)
    at IOSession.Session.executeCommand (typescript/built/local/tsserver.js:177179:33)
    at IOSession.Session.onMessage (typescript/built/local/tsserver.js:177205:35)
```

### PR DESCRIPTION
**Pull request by sheetalkamat**
Also convert one relevant test case to baseline
Fixes #50131

### CODE REVIEW

File: src/server/session.ts [hunk 1]
```diff
@@ -1868,9 +1868,9 @@ namespace ts.server {
 
         private getFileAndLanguageServiceForSyntacticOperation(args: protocol.FileRequestArgs) {
             // Since this is syntactic operation, there should always be project for the file
-            // we wouldnt have to ensure project but rather throw if we dont get project
+            // throw if we dont get project
             const file = toNormalizedPath(args.file);
-            const project = this.getProject(args.projectFileName) || this.projectService.tryGetDefaultProjectForFile(file);
+            const project = this.getProject(args.projectFileName) || this.projectService.ensureDefaultProjectForFile(file);
```
Comment (amcasey) [inline src/server/session.ts:1873]: Is this going to have perf implications?

**Review by amcasey — COMMENTED** [review]
Why would a project be required for outlining spans?  Are there project-scoped settings that affect the behavior?  Same question for (e.g.) brace matching.

**Comment by DanielRosenwasser** [comment]
So what exactly happened before? Was it something like this?

1. Open a file - a default project is automatically made.
2. Close the file - the default project is automatically closed because it has no files open.
3. Open another file - the default project is assumed not to be closed (?)

File: src/server/session.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/server/session.ts:1873]: We use it to get LanguageService To invoke the methods on .. 
There shouldnt be perf penalty as the updated function is `this.tryGetDefaultProjectForFile(fileName) || this.doEnsureDefaultProjectForFile(fileName);` so only if previously we were throwing no project found we would be doing the extra work to ensure that there is project for the script info which is intended.

**Comment by sheetalkamat** [comment]
> So what exactly happened before? 

- Open a file say src/file.ts. Opens configured project say tsconfig.json
- Close this file. Configured project is not collected because the removal of projects and scriptinfos that are orphan happens only when opening file
- open another file say tests/file.ts  It finds a tsconfig file tests/tsconfig.json but that does not include this file. But existing tsconfig.json has this file so we dont create inferred project. After this collecting orphan project collects tsconfig.json because there is no open file which will open this project (tests/file.ts does not count because any file in tests folder would open tests/tsconfig.json and not tsconfig.json) So now tests/file.ts is not part of any project
- Call outlinging span and we were throwin error before the fix. With the fix it sees there is no project for this file so it creates inferred project for this file. (just like it would have if we didnt do first two steps)

**Comment by amcasey** [comment]
I'm still missing why we'd need a project at all to handle this request.  Is that just an unfortunate consequence of our design?

File: src/testRunner/unittests/tsserver/projects.ts [hunk 2]
```diff
@@ -1620,5 +1609,40 @@ namespace ts.projectSystem {
...
+
+        it("file opened is in configured project that will be removed", () => {
+            const testsConfig: File = {
+                path: `${tscWatch.projectRoot}/playground/tsconfig.json`,
+                content: "{}"
+            };
+            const testsFile: File = {
+                path: `${tscWatch.projectRoot}/playground/tests.ts`,
+                content: `export function foo() {}`
+            };
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/tsserver/projects.ts:1621]: This is a bit of a nit - but really, is the idea that that `playground` contains some sort of implementation file, and `playground/tests` would contain the test file? In theory, you should only need 2 tsconfigs and 2 `.ts` files - but you don't need `innerSrcFile`, right?

```
src/
├── tsconfig.json
├── foo.ts
└── spec/
    ├── tsconfig.json
    └── foo.ts
```

**Comment by sheetalkamat** [comment]
> I'm still missing why we'd need a project at all to handle this request. Is that just an unfortunate consequence of our design?

Thats our design for now.. in most cases the open files are already part of the project (the scenario we are fixing is just unfortunate) so it was never priority to experiment separate project for syntax only operations. (To get hold of LS on which we can forward those requests)

**Review by amcasey — APPROVED** [review]

**Pull request merged**